File: ic14-concurrency/src/factorial.ts
/**
 * @param n must be integer >= 0
 * @returns n!
 */
export function factorial(n: number): bigint {
    let result = 1n;
    for (let i = 1n; i <= n; ++i) {
        result *= i;
    }
    console.log(`at time ${process.hrtime.bigint().toLocaleString()}: computed ${n}!`);
    return result;
}

/**
 * @param start must be integer
 * @returns increasing sequence of integers from `start` to infinity
 */
export function* rangeToInfinity(start: number): Iterable<number> {
    let n = start;

    while (true) {
        yield n++;
    }
}

/**
 * @param f
 * @param iterable
 * @returns sequence resulting from applying `f` to every value of `iterable` in order
 */
export function* map<T,U>(f: (t:T)=>U, iterable: Iterable<T>): Iterable<U> {
    for (const x of iterable) {
        yield f(x);
    }
}

/** Main program.  Computes an increasing infinite sequence of factorials starting with 9000!. */
export function main(): void {
    for (const _ of map(factorial, rangeToInfinity(9000))) {
    }
}

File: ic14-concurrency/src/factorial.py
import time

def factorial(n):
    '''
    Returns n!, where n must be an integer >= 0
    '''
    result = 1
    for i in range(1, n+1):
        result *= i
    print(f'at time {time.monotonic_ns():,}: computed {n}!')
    return result

def range_to_infinity(start):
    '''
    Yields increasing sequence of integers from `start` to infinity,
    where start must be an integer
    '''
    n = start
    while True:
        yield n
        n += 1
    # TODO

def map(f, iterable):
    '''
    Yields sequence resulting from applying `f` to every value of `iterable` in order
    '''
    for x in iterable:
        yield f(x)

if __name__ == '__main__':
    # main program:
    # compute an increasing infinite sequence of factorials starting with 9000!
    for _ in map(factorial, range_to_infinity(9000)):
        pass
